ic-identity-hsm/src/hsm.rs

```diff
@@ -0,0 +1,358 @@
+    ExpectedEcPointOctetString,
+
+    #[error("EcPoint is empty")]
+    EcPointEmpty,
+
+    #[error("Attribute with type={0} not found")]
+    AttributeNotFound(CK_ATTRIBUTE_TYPE),
+
+    #[error("Invalid EcParams.  Expected prime256v1 {:02x?}, actual is {:02x?}", .expected, .actual)]
+    InvalidEcParams { expected: Vec<u8>, actual: Vec<u8> },
+}
+
+/// An identity based on an HSM
+pub struct HardwareIdentity {
+    key_id: KeyIdVec,
+    ctx: Ctx,
+    session_handle: CK_SESSION_HANDLE,
+    logged_in: bool,
+    public_key: DerPublicKeyVec,
+}
+
+impl HardwareIdentity {
+    /// Create an identity using a specific key on an HSM.
+    /// The filename will be something like /usr/local/lib/opensc-pkcs11.s
+    /// The key_id must refer to a ECDSA key with parameters prime256v1 (secp256r1)
+    /// The key must already have been created.  You can create one with pkcs11-tool:
+    /// $ pkcs11-tool -k --slot $SLOT -d $KEY_ID --key-type EC:prime256v1 --pin $PIN
+    pub fn new<P>(
+        filename: P,
+        slot_id: CK_SLOT_ID,
+        key_id: &str,
+        pin: &str,
+    ) -> Result<HardwareIdentity, HardwareIdentityError>
+    where
+        P: AsRef<Path>,
+    {
+        let ctx = Ctx::new_and_initialize(filename)?;
+        let session_handle = open_session(&ctx, slot_id)?;
+        let logged_in = login_if_required(&ctx, session_handle, pin, slot_id)?;
+        let key_id = str_to_key_id(key_id)?;
+        let public_key = get_der_encoded_public_key(&ctx, session_handle, &key_id)?;
+
+        Ok(HardwareIdentity {
+            key_id,
+            ctx,
+            session_handle,
+            logged_in,
+            public_key,
+        })
+    }
+}
+
+impl Identity for HardwareIdentity {
+    fn sender(&self) -> Result<Principal, String> {
+        Ok(Principal::self_authenticating(&self.public_key))
+    }
+    fn sign(&self, msg: &[u8], _principal: &Principal) -> Result<Signature, String> {
+        let hash = hash_message(msg);
+        let signature = self.sign_hash(&hash)?;
+
+        Ok(Signature {
+            public_key: self.public_key.clone(),
+            signature,
+        })
+    }
+}
+
+// We open a session for the duration of the lifetime of the HardwareIdentity.
+fn open_session(
+    ctx: &Ctx,
+    slot_id: CK_SLOT_ID,
+) -> Result<CK_SESSION_HANDLE, HardwareIdentityError> {
+    let flags = CKF_SERIAL_SESSION;
+    let application = None;
+    let notify = None;
+    let session_handle = ctx.open_session(slot_id, flags, application, notify)?;
+    Ok(session_handle)
+}
+
+// We might need to log in.  This requires the PIN.
+fn login_if_required(
+    ctx: &Ctx,
+    session_handle: CK_SESSION_HANDLE,
+    pin: &str,
+    slot_id: CK_SLOT_ID,
+) -> Result<bool, HardwareIdentityError> {
+    let token_info = ctx.get_token_info(slot_id)?;
+    let login_required = token_info.flags & CKF_LOGIN_REQUIRED != 0;
+
+    if login_required {
+        ctx.login(session_handle, CKU_USER, Some(&pin))?;
```

Comment:
Does this only work on command line prompts / terminals? What if I do `cat | dfx --identity hsm call ...`, will this fail or hang forever?

Reply:
There is no prompt / hang, because the PIN is passed as a parameter.  `dfx` will get it from environment variable `DFX_HSM_PIN`.

Reply:
Ah I see. It sets the pin to the `&str` you gave. Okay. Not blocking this, but I think the flow is broken; from the PoV of this library we should ask the callee for the pin only if it's needed, but this flow requires it all the time. For dfx that makes sense, but another library could do a prompt for the PIN. So instead of always take a &str, we should take a `Fn() -> &str` and call it only when the pin is needed.

WDYT? Like I said, it's non blocking but if you want to improve.

Reply:
Sounds great, I will make this change